### PR TITLE
Change gogoproto compiler to protoc-gen-go-cast

### PR DESCRIPTION
**Pull request by 0xKiwi**
**What type of PR is this?**
This PR replaces all gogoproto compilers with our own native protobuf v2 compiler. We now support optional fields no longer rely on gogoproto for ethereumapis.

**What does this PR do? Why is it needed?**
This PR unblocks #7510 to continue implementation of previously blocked endpoints/features.

More context on changes made here: https://hackmd.io/wO5rzfVWS3GrS-L7FNXKGg

Majority of changes are:
- BUILD.bazel dependency changes
- Import changes from ptypes.Empty to emptypb.Empty

Please pay attention to changes I've commented on!

### CODE REVIEW

File: beacon-chain/gateway/gateway.go [hunk 1]
```diff
@@ -56,10 +56,11 @@ func (g *Gateway) Start() {
 	g.conn = conn
 
 	gwmux := gwruntime.NewServeMux(
-		gwruntime.WithMarshalerOption(
-			gwruntime.MIMEWildcard,
-			&gwruntime.JSONPb{OrigName: false, EmitDefaults: true},
-		),
```
Comment (0xKiwi) [inline beacon-chain/gateway/gateway.go:62]: Need help figuring out what to do here

File: beacon-chain/rpc/validator/proposer.go [hunk 2]
```diff
@@ -278,15 +280,15 @@ func (vs *Server) inRangeVotes(ctx context.Context,
...
 		}
 	}
 
 	return inRangeVotes, nil
 }
 
-func chosenEth1DataMajorityVote(votes []eth1DataSingleVote) eth1DataAggregatedVote {
-	var voteCount []eth1DataAggregatedVote
+func chosenEth1DataMajorityVote(votes []*eth1DataSingleVote) *eth1DataAggregatedVote {
+	var voteCount []*eth1DataAggregatedVote
```
Comment (0xKiwi) [inline beacon-chain/rpc/validator/proposer.go:291]: Had to make this a pointer because of protobuf v2

File: proto/testing/test.proto [hunk 3]
```diff
@@ -3,11 +3,10 @@ syntax = "proto3";
 package testing;
 
 import "google/protobuf/timestamp.proto";
-import "github.com/gogo/protobuf/gogoproto/gogo.proto";
 
 message TestMessage {
   string foo = 1;
-  string bar = 2 [(gogoproto.moretags) = "fizz:\"buzz\" spec-name:\"baz\""];
```
Comment (0xKiwi) [inline proto/testing/test.proto:10]: idk about this

File: validator/rpc/gateway/gateway.go [hunk 4]
```diff
@@ -50,10 +50,10 @@ func (g *Gateway) Start() {
...
 
 	gwmux := gwruntime.NewServeMux(
-		gwruntime.WithMarshalerOption(
-			gwruntime.MIMEWildcard,
-			&gwruntime.JSONPb{OrigName: false},
-		),
+	//gwruntime.WithMarshalerOption(
+	//	gwruntime.MIMEWildcard,
+	//	&gwruntime.JSONPb{OrigName: false},
+	//),
```
Comment (0xKiwi) [inline validator/rpc/gateway/gateway.go:56]: Also unsure here

File: beacon-chain/rpc/validator/proposer.go [hunk 5]
```diff
@@ -39,12 +39,12 @@ var eth1DataNotification bool
 const eth1dataTimeout = 2 * time.Second
 
 type eth1DataSingleVote struct {
-	eth1Data    ethpb.Eth1Data
+	eth1Data    *ethpb.Eth1Data
```
Comment (0xKiwi) [inline beacon-chain/rpc/validator/proposer.go:42]: Had to make changes here because of protobuf v2. Please confirm its ok.

File: shared/messagehandler/messagehandler.go [hunk 6]
```diff
@@ -18,7 +18,7 @@ var log = logrus.WithField("prefix", "message-handler")
 
 // SafelyHandleMessage will recover and log any panic that occurs from the
 // function argument.
-func SafelyHandleMessage(ctx context.Context, fn func(ctx context.Context, message proto.Message) error, msg proto.Message) {
+func SafelyHandleMessage(ctx context.Context, fn func(ctx context.Context, message *pubsub.Message) error, msg *pubsub.Message) {
```
Comment (0xKiwi) [inline shared/messagehandler/messagehandler.go:21]: Confirm these changes are ok please, p2p uses gogo so its a little weird splitting from them.

File: shared/sszutil/deep_equal_test.go [hunk 7]
```diff
@@ -60,27 +60,27 @@ func TestDeepEqualStructs_Unexported(t *testing.T) {
...
 	assert.Equal(t, true, sszutil.DeepEqual(fork1, fork2))
 
-	fork1 = pb.Fork{
+	fork1 = &pb.Fork{
 		PreviousVersion: []byte{123},
 		CurrentVersion:  []byte{124},
 		Epoch:           1234567890,
 	}
-	fork2 = pb.Fork{
+	fork2 = &pb.Fork{
```
Comment (0xKiwi) [inline shared/sszutil/deep_equal_test.go:71]: Also had to make this change for protobuf v2.

File: slasher/detection/detect_test.go [hunk 8]
```diff
@@ -707,16 +708,20 @@ func TestServer_MapResultsToAtts(t *testing.T) {
...
-		t.Error("Expected map:")
-		for key, value := range resultsToAtts {
-			t.Errorf("Key %#x: %d atts", key, len(value))
-			t.Errorf("%+v", value)
-		}
-		t.Error("To equal:")
-		for key, value := range expectedResultsToAtts {
-			t.Errorf("Key %#x: %d atts", key, len(value))
-			t.Errorf("%+v", value)
+	for k := range expectedResultsToAtts {
```
Comment (0xKiwi) [inline slasher/detection/detect_test.go:711]: Changed this to for loop the keys because:
- Since changing to protobuf v2, reflect.DeepEqual has problems due to the unexported fields.
- sszutil.DeepEqual doesn't support maps

This essentially means we have no simple way of checking equality on maps. Is this a problem?

File: beacon-chain/operations/attestations/kv/aggregated_test.go [hunk 9]
```diff
@@ -256,7 +256,7 @@ func TestKV_Aggregated_AggregatedAttestations(t *testing.T) {
 	sort.Slice(returned, func(i, j int) bool {
 		return returned[i].Data.Slot < returned[j].Data.Slot
 	})
-	assert.DeepEqual(t, atts, returned)
+	assert.DeepSSZEqual(t, atts, returned)
```
Comment (0xKiwi) [inline beacon-chain/operations/attestations/kv/aggregated_test.go:259]: We can no longer check protobuf arrays with reflect, so every check for protobuf arrays now goes through sszutil. Is this fine?  I don't think I've changed any runtime code.

File: beacon-chain/rpc/beacon/attestations.go [hunk 10]
```diff
@@ -196,7 +196,7 @@ func (bs *Server) ListIndexedAttestations(
 // aggregated attestations currently in the pool at the start of a slot and sends
 // them over a gRPC stream.
 func (bs *Server) StreamAttestations(
-	_ *ptypes.Empty, stream ethpb.BeaconChain_StreamAttestationsServer,
+	_ *emptypb.Empty, stream ethpb.BeaconChain_StreamAttestationsServer,
```
Comment (0xKiwi) [inline beacon-chain/rpc/beacon/attestations.go:199]: These were part of the changes to protobuf v2.

File: beacon-chain/state/stateV0/state_trie.go [hunk 11]
```diff
@@ -18,14 +19,22 @@ import (
...
-	return InitializeFromProtoUnsafe(proto.Clone(st).(*pbp2p.BeaconState))
+	return InitializeFromProtoUnsafe(hydrateState(proto.Clone(st).(*pbp2p.BeaconState)))
+}
+
+func hydrateState(st *pbp2p.BeaconState) *pbp2p.BeaconState {
+	if st.Validators == nil {
+		st.Validators = []*ethpb.Validator{}
+	}
+	return st
 }
```
Comment (0xKiwi) [inline beacon-chain/state/stateV0/state_trie.go:35]: Added this because empty arrays after performing a proto.Clone would be nil. Wanted to make sure this is a valid workaround.

File: beacon-chain/state/stateutil/state_root_test.go [hunk 12]
```diff
@@ -20,7 +19,9 @@ func TestState_FieldCount(t *testing.T) {
 	typ := reflect.TypeOf(pb.BeaconState{})
 	numFields := 0
 	for i := 0; i < typ.NumField(); i++ {
-		if strings.HasPrefix(typ.Field(i).Name, "XXX_") {
+		if typ.Field(i).Name == "state" ||
+			typ.Field(i).Name == "sizeCache" ||
+			typ.Field(i).Name == "unknownFields" {
```
Comment (0xKiwi) [inline beacon-chain/state/stateutil/state_root_test.go:24]: Replacing the gogo fields with unexported fields.

File: shared/sszutil/deep_equal.go [hunk 13]
```diff
@@ -302,12 +302,21 @@ func DeepEqual(x, y interface{}) bool {
...
 		// Exclude unexported fields for protos.
 		return deepValueEqualExportedOnly(v1, v2, make(map[visit]bool), 0)
 	}
 	return deepValueEqual(v1, v2, make(map[visit]bool), 0)
 }
+
+func IsProto(item interface{}) bool {
+	typ := reflect.TypeOf(item)
+	kind := typ.Kind()
+	if kind != reflect.Slice && kind != reflect.Array && kind != reflect.Map {
```
Comment (0xKiwi) [inline shared/sszutil/deep_equal.go:315]: Map check might not be needed here. sszutil doesnt support maps anyways

**Comment by 0xKiwi** [comment]
Todo list of changes after call with Preston:

- Add comment to DeepEqual for new protobufs 
- Change nil validators in state error
- Change gateway back to separate 
- Try making sure go-cast up to date for the old protobuf dep
- Make gocast plugin usage specific to only targets that need it
- Add back proto/testings/ more tags 
- Confirm pubsub.message string() implementation

File: beacon-chain/state/stateV0/state_trie.go [hunk 14]
```diff
@@ -31,10 +33,29 @@ var (
...
+		return cloned, nil
+	}
+	if cloned.Validators == nil {
+		cloned.Validators = []*ethpb.Validator{}
+	}
+	return cloned, nil
 }
 
-// InitializeFromProtoUnsafe directly uses the beacon state protobuf pointer
+// InitializeFromProtoUnsafe direc tly uses the beacon state protobuf pointer
```
Comment (terencechain) [inline beacon-chain/state/stateV0/state_trie.go:58]:
```suggestion
// InitializeFromProtoUnsafe directly uses the beacon state protobuf pointer
```

File: beacon-chain/state/stateV0/state_trie.go [hunk 15]
```diff
@@ -31,10 +33,29 @@ var (
...
 func InitializeFromProto(st *pbp2p.BeaconState) (*BeaconState, error) {
-	return InitializeFromProtoUnsafe(proto.Clone(st).(*pbp2p.BeaconState))
+	clonedState, err := CloneBeaconState(st)
+	if err != nil {
+		return nil, err
+	}
+	return InitializeFromProtoUnsafe(clonedState)
+}
+
+// CloneBeaconState --
```
Comment (terencechain) [inline beacon-chain/state/stateV0/state_trie.go:43]: Full comment?

File: beacon-chain/cache/depositcache/pending_deposits_test.go [hunk 16]
```diff
@@ -90,7 +90,7 @@ func TestPendingDeposits_OK(t *testing.T) {
 		{Proof: [][]byte{[]byte("A")}},
 		{Proof: [][]byte{[]byte("B")}},
 	}
-	assert.DeepEqual(t, expected, deposits)
+	assert.DeepSSZEqual(t, expected, deposits)
```
Comment (nisdas) [inline beacon-chain/cache/depositcache/pending_deposits_test.go:93]: why is this changed ?

File: beacon-chain/operations/attestations/kv/aggregated_test.go (hunk 9 above)
Comment (nisdas) [reply beacon-chain/operations/attestations/kv/aggregated_test.go:259]: can you explain why reflect can't catch this ? The underlying data structure should be the same here.

File: beacon-chain/p2p/encoder/ssz_test.go [hunk 17]
```diff
@@ -8,13 +8,14 @@ import (
 	"math"
 	"testing"
 
-	"github.com/gogo/protobuf/proto"
+	gogo "github.com/gogo/protobuf/proto"
```
Comment (nisdas) [inline beacon-chain/p2p/encoder/ssz_test.go:11]: why use gogo here ? aren't we trying to remove it

File: beacon-chain/rpc/beaconv1/blocks.go [hunk 18]
```diff
@@ -70,11 +70,11 @@ func (bs *Server) ListBlockHeaders(ctx context.Context, req *ethpb.BlockHeadersR
 			return nil, status.Errorf(codes.Internal, "Could not retrieve blocks: %v", err)
 		}
 	} else {
-		_, blks, err = bs.BeaconDB.BlocksBySlot(ctx, req.Slot)
+		_, blks, err = bs.BeaconDB.BlocksBySlot(ctx, *req.Slot)
```
Comment (nisdas) [inline beacon-chain/rpc/beaconv1/blocks.go:73]: why are you providing a pointer here

File: beacon-chain/rpc/validator/proposer.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:43]: but why ? Can you clarify why this needs to be a pointer

File: beacon-chain/rpc/validator/proposer.go [hunk 19]
```diff
@@ -40,12 +40,12 @@ var eth1DataNotification bool
...
 	blockHeight *big.Int
 }
 
 type eth1DataAggregatedVote struct {
-	data  eth1DataSingleVote
+	data  *eth1DataSingleVote
```
Comment (nisdas) [inline beacon-chain/rpc/validator/proposer.go:48]: same here

File: beacon-chain/state/stateV0/state_trie.go [hunk 20]
```diff
@@ -31,7 +32,31 @@ var (
...
+		return nil, errors.New("received nil state")
+	}
+	cloned, ok := proto.Clone(st).(*pbp2p.BeaconState)
+	if !ok {
+		return nil, errors.New("not a beacon state")
+	}
+	if cloned == nil {
+		return nil, errors.New("beacon state clone is empty")
+	}
+	if cloned.Validators == nil {
```
Comment (nisdas) [inline beacon-chain/state/stateV0/state_trie.go:56]: why only the validators field. It could apply to many other fields

File: fuzz/block_fuzz.go [hunk 21]
```diff
@@ -191,7 +191,7 @@ func BeaconFuzzBlock(b []byte) {
 		return
 	}
 
-	if err := s.FuzzBeaconBlockSubscriber(ctx, msg); err != nil {
+	if err := s.FuzzBeaconBlockSubscriber(ctx, input.Block); err != nil {
```
Comment (nisdas) [inline fuzz/block_fuzz.go:194]: why is being changed, `msg` and `input.Block` arent the same.

**Review by nisdas — COMMENTED** [review]
Did a first pass through everything, had some thoughts on them

File: fuzz/block_fuzz.go (hunk 21 above)
Comment (0xKiwi) [reply fuzz/block_fuzz.go:194]: Would appreciate more guidance on this. Libp2p uses gogoproto which complicated some parts here

File: beacon-chain/state/stateV0/state_trie.go (hunk 20 above)
Comment (0xKiwi) [reply beacon-chain/state/stateV0/state_trie.go:56]: Haven't encountered problems with others but will add more cases.

File: beacon-chain/rpc/validator/proposer.go (hunk 19 above)
Comment (0xKiwi) [reply beacon-chain/rpc/validator/proposer.go:48]: This was due to a change of how protobuf handles structs not being a pointer. If this weren't a pointer protobuf would it would fail compilation.

File: beacon-chain/rpc/beaconv1/blocks.go (hunk 18 above)
Comment (0xKiwi) [reply beacon-chain/rpc/beaconv1/blocks.go]: Because this is now an optional field, protobuf handles it as a pointer, I'm dereferencing it to use it. I've changed the logic to handle the optional field properly now.

File: beacon-chain/p2p/encoder/ssz_test.go (hunk 17 above)
Comment (0xKiwi) [reply beacon-chain/p2p/encoder/ssz_test.go:11]: This uses a function `gogo.EncodeVarint`, which I could not find in the standard proto library. Even if this uses gogo, its not as crucial as anything related to the APIs. Can look into finding a more native alternative for this in a future PR.

File: beacon-chain/cache/depositcache/pending_deposits_test.go (hunk 16 above)
Comment (0xKiwi) [reply beacon-chain/cache/depositcache/pending_deposits_test.go:93]: Reflect no longer works for v2 protobuf structs because of unexported fields. And this is an array of proto messages, so the only equality check the works  really is DeepSSZEqual.

File: fuzz/block_fuzz.go (hunk 21 above)
Comment (0xKiwi) [reply fuzz/block_fuzz.go:194]: So this issue came from our fuzz_exports.go changing from taking in a gogo.Message to a proto.Message, and libp2p is gogo messages, so not exactly sure how to handle this. Maybe we can keep using gogo for libp2p things and convert from there?

Comment (0xKiwi) [reply fuzz/block_fuzz.go]: Ok, looked at it again, undid these changes to keep results the same. Its fine if it uses gogo here, as this is code pertaining to libp2p, not our protobufs.

Comment (0xKiwi) [reply fuzz/block_fuzz.go]: Encountering problems though, would appreciate some assistance.

File: beacon-chain/state/stateV0/state_trie.go (hunk 20 above)
Comment (nisdas) [reply beacon-chain/state/stateV0/state_trie.go:56]: Did this only come up in tests ?

File: beacon-chain/rpc/validator/proposer.go (hunk 19 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:48]: Not sure I understand, what does this have to do with protobufs ? its a normal struct so any protobuf specific gotchas shouldnt apply here.

File: beacon-chain/state/stateV0/state_trie.go (hunk 20 above)
Comment (0xKiwi) [reply beacon-chain/state/stateV0/state_trie.go:56]: I think so, yes, let me confirm if it arises in runtime

File: fuzz/block_fuzz.go (hunk 21 above)
Comment (prestonvanloon) [reply fuzz/block_fuzz.go:194]: @0xKiwi nice fix here. The FuzzBeaconBlockSubscriber was expecting a proto.Message and we were providing the incorrect message. Thanks!

File: beacon-chain/p2p/encoder/ssz_test.go (hunk 17 above)
Comment (prestonvanloon) [reply beacon-chain/p2p/encoder/ssz_test.go:11]: It's here: https://github.com/golang/protobuf/blob/ae97035608a719c7a1c1c41bed0ae0744bdb0c6f/proto/buffer.go#L26

Let's remove gogo as much as possible, it is not maintained.

File: go.mod [hunk 22]
```diff
@@ -131,3 +130,5 @@ require (
 replace github.com/ethereum/go-ethereum => github.com/prysmaticlabs/bazel-go-ethereum v0.0.0-20210420143944-f4dfc9744288
 
 replace github.com/json-iterator/go => github.com/prestonvanloon/go v1.1.7-0.20190722034630-4f2e55fcf87b
+
+replace github.com/grpc-ecosystem/grpc-gateway/v2 => github.com/prysmaticlabs/grpc-gateway/v2 v2.3.1-0.20210330221950-3beba9ed0e19
```
Comment (prestonvanloon) [inline go.mod:134]:
```suggestion
// See https://github.com/prysmaticlabs/grpc-gateway/issues/2
replace github.com/grpc-ecosystem/grpc-gateway/v2 => github.com/prysmaticlabs/grpc-gateway/v2 v2.3.1-0.20210330221950-3beba9ed0e19
```

File: proto/testing/test.proto (hunk 3 above)
Comment (prestonvanloon) [reply proto/testing/test.proto:10]: Discussed offline a while ago. It's not a critical feature, your change is sufficient

File: proto/validator/accounts/v2/BUILD.bazel [hunk 23]
```diff
@@ -13,53 +13,57 @@ proto_library(
...
+    visibility = ["//visibility:private"],
     deps = [
         "//proto/beacon/rpc/v1:go_default_library",
         "@com_github_prysmaticlabs_eth2_types//:go_default_library",
+        "@io_bazel_rules_go//proto/wkt:descriptor_go_proto",
+        #"@com_github_golang_protobuf//descriptor:go_default_library",
+        "@com_github_golang_protobuf//proto:go_default_library",
+        "@io_bazel_rules_go//proto/wkt:empty_go_proto",
+        #"@io_bazel_rules_go//proto/wkt:empty_go_proto",
         "@com_github_prysmaticlabs_ethereumapis//eth/v1alpha1:go_default_library",
```
Comment (prestonvanloon) [inline proto/validator/accounts/v2/BUILD.bazel:58]: Can you clean this please? Some commented stuff / duplicates.

File: beacon-chain/rpc/beaconv1/blocks.go [hunk 24]
```diff
@@ -70,11 +70,15 @@ func (bs *Server) ListBlockHeaders(ctx context.Context, req *ethpb.BlockHeadersR
...
+		slot := bs.ChainInfoFetcher.HeadSlot()
+		if req.Slot != nil {
+			slot = *req.Slot
+		}
+		_, blks, err = bs.BeaconDB.BlocksBySlot(ctx, slot)
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for slot %d: %v", req.Slot, err)
 		}
-		_, blkRoots, err = bs.BeaconDB.BlockRootsBySlot(ctx, req.Slot)
+		_, blkRoots, err = bs.BeaconDB.BlockRootsBySlot(ctx, slot)
```
Comment (prestonvanloon) [inline beacon-chain/rpc/beaconv1/blocks.go:81]: This looks like a logic change. Can you explain?

File: beacon-chain/rpc/validator/proposer.go (hunk 5 above)
Comment (prestonvanloon) [reply beacon-chain/rpc/validator/proposer.go:43]: Please answer that in a reply here, if possible. Even a playground link would be enough to communicate the error or issue requiring this change.

File: beacon-chain/state/stateV0/state_trie.go [hunk 25]
```diff
@@ -31,7 +32,40 @@ var (
...
+	if err != nil {
+		return nil, err
+	}
+	return InitializeFromProtoUnsafe(clonedState)
+}
+
+// CloneBeaconState serves as a wrapper for proto.Clone to apply any
+// post-cloning changes needed. This was added due to a change in protobuf v2
+// which encodes empty slices as nil.
+func CloneBeaconState(st *pbp2p.BeaconState) (*pbp2p.BeaconState, error) {
```
Comment (prestonvanloon) [inline beacon-chain/state/stateV0/state_trie.go:45]: Is this sitll necessary? I thought that SszDeepEqual fixed this issue?

File: beacon-chain/rpc/validator/proposer.go (hunk 19 above)
Comment (0xKiwi) [reply beacon-chain/rpc/validator/proposer.go]: I replicated the problem but it turns out this change was not needed, thank you for pointing it out!

File: beacon-chain/state/stateV0/state_trie.go (hunk 25 above)
Comment (0xKiwi) [reply beacon-chain/state/stateV0/state_trie.go:45]: Testing would fail with `nil validators in state`. Maybe its a problem that can be solved without this change. Let me see.

File: beacon-chain/rpc/beaconv1/blocks.go (hunk 24 above)
Comment (0xKiwi) [reply beacon-chain/rpc/beaconv1/blocks.go:81]: Since req.Slot is now a pointer, changing the api endpoint to handle it properly for now.

File: beacon-chain/rpc/validator/proposer.go (hunk 5 above)
Comment (0xKiwi) [reply beacon-chain/rpc/validator/proposer.go:43]:
```
beacon-chain/rpc/validator/proposer.go:314:18: assignment copies lock value to currentVote: github.com/prysmaticlabs/prysm/beacon-chain/rpc/validator.eth1DataAggregatedVote contains github.com/prysmaticlabs/prysm/beacon-chain/rpc/validator.eth1DataSingleVote contains github.com/prysmaticlabs/ethereumapis/eth/v1alpha1.Eth1Data contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex (copylocks)
beacon-chain/rpc/validator/proposer.go:318:9: return copies lock value: github.com/prysmaticlabs/prysm/beacon-chain/rpc/validator.eth1DataAggregatedVote contains github.com/prysmaticlabs/prysm/beacon-chain/rpc/validator.eth1DataSingleVote contains github.com/prysmaticlabs/ethereumapis/eth/v1alpha1.Eth1Data contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex (copylocks)
```

I would encounter these errors if it were not a pointer.

File: beacon-chain/p2p/encoder/ssz_test.go (hunk 17 above)
Comment (0xKiwi) [reply beacon-chain/p2p/encoder/ssz_test.go:11]: Thats protobuf v1, but maybe its good since its a switch away from gogo, thanks

Comment (nisdas) [reply beacon-chain/p2p/encoder/ssz_test.go:11]: seems like you are still using gogo ?

File: beacon-chain/rpc/beaconv1/blocks.go (hunk 24 above)
Comment (nisdas) [reply beacon-chain/rpc/beaconv1/blocks.go:81]: This doesn't make any sense though, you now only fetch from the head slot which renders any data in the request pointless. Whatever a user requests, they will still get the same data.

File: beacon-chain/rpc/validator/proposer.go (hunk 5 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:43]: This fails our static analyzer ( copyLocks), we can either add an exception for the analyzer here or actually find a way to fix this. This is a pretty bizarre restriction, we should be able to specify proto objects by value instead of only reference.

Comment (nisdas) [reply beacon-chain/rpc/validator/proposer.go:43]: Ok looking at this, seems its not allowed to be copied by value:
https://github.com/protocolbuffers/protobuf-go/blob/v1.26.0/internal/impl/message_reflect.go#L363

while annoying it should be fine to leave it as a pointer.

File: beacon-chain/p2p/encoder/ssz_test.go (hunk 17 above)
Comment (0xKiwi) [reply beacon-chain/p2p/encoder/ssz_test.go:11]: I tried the change but it led to test failures here:
https://buildkite.com/prysmatic-labs/prysm/builds/50217

Not sure why....

File: beacon-chain/state/stateV0/state_trie.go (hunk 20 above)
Comment (0xKiwi) [reply beacon-chain/state/stateV0/state_trie.go]: Confirmed, only in testing, got rid of these changes.

File: beacon-chain/rpc/beaconv1/blocks.go (hunk 24 above)
Comment (0xKiwi) [reply beacon-chain/rpc/beaconv1/blocks.go:81]: Wdym? I check if  the request they sent is nil and then pass in that info instead. Why would it always be head slot?

File: beacon-chain/p2p/encoder/ssz_test.go (hunk 17 above)
Comment (0xKiwi) [reply beacon-chain/p2p/encoder/ssz_test.go:11]: Spoke to preston, we can address this in a future PR. Would rather this PR just focus on API related stuff.

File: beacon-chain/rpc/beaconv1/blocks.go (hunk 24 above)
Comment (nisdas) [reply beacon-chain/rpc/beaconv1/blocks.go:81]: Nvm ignore the comment, was looking at something else and got it mixed with this.

**Review by nisdas — COMMENTED** [review]
The PR looks good to me, will run it for a while on my end to make sure there are no other surprises

**Review by nisdas — APPROVED** [review]
LGTM

**Review by rauljordan — APPROVED** [review]
